### DISCUSSION TITLE
Quick question

### DISCUSSION THREAD

**Discussion by karolyi** [LetsDNS/letsdns#5] [General]
Hey,

with this project, you might just have spared me of implementing something similar in python :) I already started putting something together but then I fiddled around a bit on pypi wondering if someone already did it, and found this project. I don't necessarily want to reinvent the wheel.

My scenario is a FreeBSD box with nginx+letsencrypt running in one jail, and several other daemons running in separate jails: coturn, postfix+dovecot, prosody and the likes.

The certificate directories for the daemons are null-mounted (same as bind mounts on linux), and I've created some scripts to run from cron and restart/reload their respective daemons within 24hrs when a newer certificate is present.

I've started to read the documentation and it outlined exactly this situation with using key rollovers.

There was one last thing missing with the rollover from the documentation though: when the daemons are restarted in 24hrs upon certificate update, does your project remove the old TLSA records after another 24hrs?

Also I assume that the TLSA records themselves will have a 24hr TTL. I am right here? It only makes sense.

Cheers,
László

**Comment by karolyi** [comment]
Hm, after more reading, I've got a bit confused.

If a proper key rollover is to be made, the following should be the workflow, given the TTL is 1 day:

- Dehydrated/certbot gets a newly signed certificate
- this program immediately places a TLSA record _next to_ the original one
- wait for 24 hours to pass for caches to expire, then replace the certificate
- after another 24 hrs, it is assumed that the respective daemons have reloaded and are using the new certificate
- at which point, the 'old' TLSA record can be cleared.

is this something this program can achieve?

**Comment by rseichter** [comment]
Short answer: LetsDNS removes old records. If you look at the source code for the action classes (e.g. `DnsLiveUpdate`), you will see that each includes a cleanup step which removes existing DNS records. If you want to see what is happening exactly, you can configure `LOG_LEVEL=DEBUG` or define a section which uses `action=nsupdate-stdout`.

**Comment by karolyi** [comment]
I hope to be able to look into this today, and see if your rollover method fulfills my needs.

I still use certbot with my certs though (with the nginx plugin) so I'll also need to figure out a way to a proper transition.

**Comment by rseichter** [comment]
Please note that the documentation mentions that DNS record TTL is assumed to be no longer than one hour. There could be a chain of caching DNS resolvers between you and whoever is querying the TLSA records, plus some pre-fetching thrown in mix. Longer TTL might work fine, but the risk of possible caching issues on third-party resolvers increases with the TTL. I use 3600 seconds, for myself and my customers, and I never experienced any problems.

**Comment by karolyi** [comment]
Hey,

I'm looking into the code now. Yeah, by the looks of it, you only maintain 1 TLSA record at any time, and you delete any other existing ones. This is not what I want actually, but there's a lot of code/inspiration I can reuse in my project, given I'm not gonna make a PR here to change that behavior (optionally).

Secondly, a caching recursive resolver should only store a cached record for its TTL, except if it's configured to serve expired records (see unbound's [`serve-expired`](https://unbound.docs.nlnetlabs.nl/en/latest/manpages/unbound.conf.html#unbound-conf-serve-expired) options.I think if your server is unreachable, you'll have bigger problems than the TLSA records being invalid, due to DNSSEC. With that, my original workflow of replacing TLSA records should be valid. The whole safe key rollover mechanism is done with keeping this TTL expiration in mind.

So the question is: are you willing to change the code to suit my needs, in which case I'll try to prepare a PR when time allows, or should I just make my own version of it?

**Comment by rseichter** [comment]
LetsDNS does not "only maintain 1 TLSA record at any time". How did you arrive at that conclusion? As for TTL: If you want higher values, nobody is stopping you, just space out the calls to LD and your certificate distribution accordingly.

**Comment by karolyi** [comment]
I might be wrong, but this line seems to delete any existing records before setting another one with the same name:

https://github.com/LetsDNS/letsdns/blob/f47166be9c1cd3cae0065ee096a2f6144f5d397b/letsdns/liveupdate.py#L52

So I assumed that's the case.

**Comment by rseichter** [comment]
That `delete` statement doesn't stand alone, now does it? 😉 The same function call also adds 1..n new records, as defined in the configuration file.

**Comment by karolyi** [comment]
So if I understand it right (please correct me if I dont), you generate TLSA records for the available (file readable) certificates in the configuration, and remove the ones that aren't available anymore?

So if the file in `cert_queued_path` gets removed, the TLSA record gets removed for it as well?

**Comment by rseichter** [comment]
Think of it this way: Each time LetsDNS is run, it takes stock of available certificates by scanning all _cert_whatever_path_ entries, one configuration section at a time. If certificate files exist at these locations, hashes are calculated, otherwise the locations are ignored for the current run. Once data collection is complete, LetsDNS removes all existing TLSA records and immediately writes new records based on the hashes, de facto replacing the TLSA records for the configured domain. In other words, LetsDNS writes a representation of currently existing certificate information (i.e. at the time of running) to DNS, overwriting whatever may have previously been stored in DNS.

**Comment by karolyi** [comment]
I see, thanks for the clarification. I'll look into if I can achieve my wished functionality with it.

**Comment by rseichter** [comment]
@karolyi Have you had the time for further tests? I am curious.

**Comment by karolyi** [comment]
Hey,

upon looking into what this tool does (and how it does it), and since I'm a python guy too, I've decided to create my own that will suit my needs and also is a more bundled solution: https://gitea.ksol.io/karolyi/daneupdate

It's still in pre-alpha I'd say, but I have a clear vision as to where I want to go with it, and it will suit my needs without any external (non-python) dependencies such as an acme client that you use.

"More bundled" means, It basically hooks into certbot internals and aims to completely replace the `renew` command, while working with already deployed certificates and upholding the TLSA record rollovers, outlined above by me. The upshot is, by virtue of hooking into certbot, it'll automatically discover and deal with nginx or any other plugins that are used with certbot.

At this point I intend to use the DNS UPDATE method, so no nsupdate/hetzner API support. Maybe at later stage but like I said, it is currently tailored to my needs so your tool might be more appropriate for those who have those requirements. I don't plan on competing. I don't think it'll end up on pypi because pypi is anti free-speech (they don't like my license, lol), and in terms of documentation, yours outcompetes mine any given day, which is okay since it's for me :) 

Looking into your python code, as a guy who does python professionally since 2002, I must say it's clean and concise, so kudos for that.

**Reply by rseichter** [reply]
Thanks for satisfying my curiosity. Yours is a different approach than mine, and it will be interesting to see where it leads you. I wish you luck in your endeavour.

**Comment by karolyi** [comment]
Hey,

just a quick update, the tool started working in my environment :) Docs still missing but it does the job the way I wished.

**Comment by rseichter** [comment]
Nice. It always feels good when a self-written software project comes to fruition, doesn't it?

**Comment by karolyi** [comment]
It is, have done it many times before. :)

I'll put some links to your project here into the documentation.

**Comment by rseichter** [comment]
Cheers, I appreciate that.
